File: mobile/lib/features/tickers/data/ticker.dart
import '../../settings/data/exchange_type.dart';

/// Espeja `TickerRead` (`app/schemas/ticker.py`) — una entrada del catálogo de acciones que el
/// backend sincroniza desde Polygon (`scripts/sync_tickers.py`).
class Ticker {
  const Ticker({
    required this.symbol,
    required this.name,
    required this.exchange,
    required this.primaryExchange,
    required this.assetType,
  });

  factory Ticker.fromJson(Map<String, dynamic> json) => Ticker(
        symbol: json['symbol'] as String,
        name: json['name'] as String,
        // `exchangeTypeFromWire` devuelve null para `OTHER` (el enum del cliente solo tiene
        // NASDAQ/NYSE, las dos bolsas que el producto ofrece elegir) — no es un error de
        // parseo: es un ticker de una bolsa que todavía no está en el selector.
        exchange: exchangeTypeFromWire(json['exchange'] as String?),
        primaryExchange: json['primary_exchange'] as String?,
        assetType: json['asset_type'] as String?,
      );

  final String symbol;
  final String name;

  /// `null` cuando el backend devolvió `OTHER` — ver `fromJson`.
  final ExchangeType? exchange;

  /// Código MIC crudo del proveedor (`XNAS`, `XNYS`, `ARCX`…). Se expone para poder mostrar
  /// de qué bolsa se trata cuando `exchange` es `null`, en vez de un genérico "otra".
  final String? primaryExchange;

  /// Tipo de instrumento según Polygon (`CS`, `ETF`, `ADRC`…), no el AssetType del producto.
  final String? assetType;

  /// Texto para el badge de bolsa. Cae al MIC crudo antes que a un placeholder: para un ETF de
  /// NYSE Arca es más informativo ver "ARCX" que "OTRA".
  String get exchangeLabel =>
      exchange?.displayName ?? primaryExchange ?? 'OTRA';
}

/// Espeja `TickerPage` — el catálogo es grande (~10k símbolos), así que siempre viene paginado.
class TickerPage {
  const TickerPage({required this.items, required this.total});

  factory TickerPage.fromJson(Map<String, dynamic> json) => TickerPage(
        items: (json['items'] as List)
            .map((item) => Ticker.fromJson(item as Map<String, dynamic>))
            .toList(),
        total: json['total'] as int,
      );

  final List<Ticker> items;
  final int total;
}

File: app/schemas/ticker.py
"""Schemas Pydantic v2 de `GET /api/v1/tickers` (catálogo de acciones)."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.models.enums import ExchangeType


class TickerRead(BaseModel):
    model_config = ConfigDict(strict=True, extra="forbid", from_attributes=True)

    symbol: str
    name: str
    # Se expone el MIC crudo además de la bolsa normalizada: si el cliente ve `OTHER` puede
    # entender por qué (qué bolsa real es) sin tener que consultar el backend.
    primary_exchange: str | None
    exchange: ExchangeType
    asset_type: str | None
    active: bool
    updated_at: datetime


class TickerPage(BaseModel):
    model_config = ConfigDict(strict=True, extra="forbid")

    items: list[TickerRead]
    total: int
    limit: int
    offset: int
